### PR TITLE
feat: Implement experimental AWS ECS resource attributes

### PR DESCRIPTION
**Pull request by mmanciop**
## Which problem is this PR solving?

- The resource attributes for AWS ECS returned by the AwsEcsDetector are missing all the fun stuff like ARNs, LaunchType and references to the TaskDefinition (a.k.a. "Family").

## Short description of the changes

- Implement the experimental AWS ECS resource attributes [1] using the Metadata v4 Endpoint.

## Checklist

- [ ] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.

[1] https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/cloud_provider/aws/ecs/

### CODE REVIEW

**Comment by NathanielRN** [comment]
Tagging @vasireddy99 and @bryan-aguilar to review this PR for it's changes to the AWS Resource Detectors.

**Comment by mmanciop** [comment]
Hi @vasireddy99 and @bryan-aguilar, is there any news on this? We at Lumigo would love more fine-grained info about ECS Tasks to become available, and we'd be willing to do similar contributions for other SDKs.

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:47]: No need for `async` keyword if you aren't using `await`, it just results in unnecessary code generation

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:49]: There should be a type for the response in the `http` module

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:51]: It is more typical to reject with an `Error` instead of a string since the catch clause often expects an error.

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:60]: If you look at the event ordering for http requests [1] I think it is possible for the response to not `end` but instead `error` in which case this promise will never resolve or reject. Look at the section that says "In the case of a premature connection close after the response is received"

[1]: https://nodejs.org/api/http.html#httprequesturl-options-callback

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:47]: other private methods usually get `_` prefix  and go at the end

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (dyladan) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:176]: style: missing newline at end of file

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in a034f63a4dcb878cf1c6e14ab7555b3c6186a4cf

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in 5a15762f, also added a test that was missing

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in ab0ad4f5

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in 39af2789

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: I thought this would be covered by the `request.on('error', reject);` a few lines below @dyladan

Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Actually no, you are right, brb

Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in 30b8708b

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in b04a99f8

**Comment by dyladan** [comment]
LGTM. waiting on component owner review from @NathanielRN or @willarmiros

**Comment by dyladan** [comment]
Ping  @NathanielRN or @willarmiros

**Comment by willarmiros** [comment]
Sorry for the delay on this, will take a look within the next week

**Comment by mmanciop** [comment]
There's discussion on https://github.com/open-telemetry/opentelemetry-java/pull/4574#issuecomment-1213233736 whether the ARNs like the one this PR builds are actually correct. Until we are use, let's not merge :D

**Comment by mmanciop** [comment]
**Update:** The uncertainty about ARNs seems sorted out

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (willarmiros) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:102]: Add type?

**Review by willarmiros — APPROVED** [review]
Tested this locally with the unit test's mocked TMDE response and the resource data looked good! Should be good to go once the CI merges.

**Comment by mmanciop** [comment]
@dyladan care to give it another look? Support for `aws.logs` was added since you last did IIRC.

**Comment by dyladan** [comment]
@willarmiros is the component owner this is OK to merge when he gives the 👍

**Review by willarmiros — APPROVED** [review]
lgtm, I think just the `error` needs a type to fix the linter :)

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:102]: It actually seems that, rather than interpolating the message, the best practice is to pass the error as the second argument of the logger call. Done in ea3ea2ca.

**Comment by vsakaram** [comment]
@dyladan - tagging to request your review/approval of this PR change

**Comment by dyladan** [comment]
Ah sorry I missed will's comment. i'll run the tests and merge this

**Review by dyladan — APPROVED** [review]

**Comment by dyladan** [comment]
Please update the PR as an outdated PR cannot be merged. The tests are failing for an unrelated reason that I believe has been fixed since the last push on this PR.

**Comment by Rauno56** [comment]
Would love to merge this, @mmanciop. We cannot merge out-of-date branch. Please update.

It's best to allow changes from maintainers, so we can keep it up to date ourselves. That permission doesn't work with org-owned forks though.

**Comment by mmanciop** [comment]
@Rauno56 another branch update has happened :-)

**Comment by Rauno56** [comment]
Cool... Thanks, I'll keep an eye on this and merge once CI greens.

Please make future PRs from forks under your own account to make the process a little bit easier for us! :pray:

**Comment by Rauno56** [comment]
Do you think you could increase coverage across the diff at all?

**Comment by mmanciop** [comment]
@Rauno56 I’ll give it a shot

**Comment by mmanciop** [comment]
@Rauno56 I took a page off the AmazonEks detector and used `nock` as HTTP mock server, and that has helped a bunch according to my local test runs :-)

```
-------------------------|---------|----------|---------|---------|-------------------
File                     | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-------------------------|---------|----------|---------|---------|-------------------
All files                |   94.82 |    89.06 |   92.85 |   94.75 |                   
 AwsBeanstalkDetector.ts |   95.23 |       50 |     100 |   95.23 | 52                
 AwsEc2Detector.ts       |   97.67 |      100 |     100 |   97.61 | 140               
 AwsEcsDetector.ts       |   94.66 |    90.62 |   85.71 |   94.59 | 188-189,196,223   
 AwsEksDetector.ts       |   91.54 |      100 |    92.3 |   91.42 | 142-151,213       
 AwsLambdaDetector.ts    |     100 |     62.5 |     100 |     100 | 51-58             
 index.ts                |     100 |      100 |     100 |     100 |                   
-------------------------|---------|----------|---------|---------|-------------------
```

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (Rauno56) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:67]: Are you sure this is meant to be left like that?

Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:67]: No it’s not, good catch :-)

Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in 2f79d48a9ecf97948a158065ac9ed3619cee9a5e

**Review by Rauno56 — APPROVED** [review]
Thanks for the contribution!

**Comment by mmanciop** [comment]
@NathanielRN would you please have a look?

**Comment by willarmiros** [comment]
@NathanielRN is no longer an approver - he should be removed from the CODEOWNERS on this repo. Either way we should not block on his review, this is good to be merged from my perspective.

**Review by NathanielRN — APPROVED** [review]
Approving so no one blocks on my review!

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (blumamir) [inline detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:174]: I am not very familiar with these properties, but I see they are being set as arrays with one value. Are they always expected to have just one item? can there be more?

**Review by blumamir — COMMENTED** [review]
Sorry for adding this review so late.
Added a few comments, some of them are old issues that I saw during the review and are here mostly for discussion and not to fix as part of this PR.

None is blocker but could be nice to handle if you are up for it

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:69]: Or even have the "_getMetadataV4Resource" function return [ecsResource, awsLogsResource] and merge everything in one go at one place?

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:174]: It could be more if the ECS task had multiple containers, and the tracer was reporting about all of them. This is more the kind of thing that the OpenTelemetry Collector would do though.

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:87]: IMO, no. It will break existing users out there, and every ECS detector needs a Docker detector. It strikes me as bad dev experience.

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:104]: Honestly, I do not know. All ECS detectors I found across the various SDKs that implement them, came with this behavior before I set out to add to them. It seems to me that ECS seems to set the container name as hostname rather consistently, both in EC2 and Fargate. It could be overridden with something more precise if the metadata endpoint is found (the container name comes in with that too) by parsing the container ARN, but it is not something I would do in this PR.

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:105]: It is done consistently like this in the various ECS detectors across the SDKs that have them. Not sure if it is an especially good practice tbh.

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: Done in ee70af15 (I think)

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: ee70af15

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: ee70af15

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (mmanciop) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts]: If ECS Metadata V4 changes format, half the world will break :P

I think we can trust AWS to use strict versioning for their Metadata service, as they did in the past (we are at the 4th revision, after all), and we explicitly check for the EndpointV4 metadata, which has a different environment variable name than V3.

TL;DR: I think we are going to be fine :-)

**Comment by mmanciop** [comment]
@blumamir another round?

File: detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts
Comment (blumamir) [reply detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts:94]: Another alternative is to just invoke the detect for `docker` or `container` detector and return the merged result.
It's unconfortable to me to see this code duplication, but anyway, not for this PR :)

**Review by blumamir — APPROVED** [review]

**Comment by mmanciop** [comment]
@blumamir done

**Pull request merged**
